### ISSUE TITLE
Failed to find module 'sg'

### ISSUE THREAD

**Issue by mabod** [dracutdevs/dracut#1099]
I am using EndeavourOS (Arch). With custom kernel linux-xanmod I get an error message in the journal during boot:

`systemd-modules-load[297]: Failed to find module 'sg'`

This does not happen with other kernels because the xanmod kernel does not come with this module. It is compiled into the kernel:

```
#
# SCSI support type (disk, tape, CD-ROM)
#
CONFIG_BLK_DEV_SD=y
CONFIG_CHR_DEV_ST=m
CONFIG_BLK_DEV_SR=y
CONFIG_CHR_DEV_SG=y
CONFIG_CHR_DEV_SCH=m
CONFIG_SCSI_ENCLOSURE=m
CONFIG_SCSI_CONSTANTS=y
CONFIG_SCSI_LOGGING=y
CONFIG_SCSI_SCAN_ASYNC=y
```

dracut seems to have the sg module dependency hardcoded into the `90kernel-modules` module:
```

/usr/lib/dracut/modules.d/90kernel-modules/module-setup.sh:

[...]
        # if not on hostonly mode, or there are hostonly block device
        # install block drivers
        if ! [[ $hostonly ]] || \
            for_each_host_dev_and_slaves_all record_block_dev_drv;
        then
            hostonly='' instmods sg sr_mod sd_mod scsi_dh ata_piix

            if [[ "$hostonly_mode" == "strict" ]]; then
                install_block_modules_strict
            else
                install_block_modules
            fi
        fi
[...]
```

The PC is still booting and I do not experience any misbehavior. But the message in the log is annoying. 

Can the dependency on the `sg.ko` module be removed from dracut?

**Comment by haraldh** [comment]
Is the module mentioned in `/lib/modules/*/modules.builtin` ?

**Comment by haraldh** [comment]
Did you forget to do a `depmod` ?

**Comment by mabod** [comment]
> Is the module mentioned in /lib/modules/<KERNEL_VERSION>/modules.builtin ?

```
6# grep -i sg /lib/modules/5.10.16-xanmod1-cacule-1-cacule/modules.builtin
kernel/block/bsg.ko
kernel/drivers/scsi/sg.ko
```

> Did you forget to do a depmod -a <KERNEL_VERSION> ?

No. This is done automatically when the kernel is installed.

**Comment by haraldh** [comment]
Ok, then I'll have to revisit this. Thanks for reporting!

**Comment by mabod** [comment]
This is still open and it happens with most recent xanmod kernel and most recent dracut on Arch.

I believe the issue is that this script `/usr/lib/dracut/modules.d/90kernel-modules/module-setup.sh` is trying to load module `sg.ko` even if it is built into the kernel like with the xanmod kernels.

**Comment by haraldh** [comment]
What is your output of:
```shell
$ egrep -r . /usr/lib/modules-load.d /etc/modules-load.d
```

**Comment by haraldh** [comment]
```shell
$ sudo /lib/systemd/systemd-modules-load 
Module 'msr' is built in
Module 'sg' is built in
```

**Comment by mabod** [comment]
```
27# egrep -r . /usr/lib/modules-load.d /etc/modules-load.d
/usr/lib/modules-load.d/cdrecord.conf:sg
/usr/lib/modules-load.d/pkcs8.conf:# When distributions use CONFIG_PKCS8_PRIVATE_KEY_PARSER=m kernel option,
/usr/lib/modules-load.d/pkcs8.conf:# using keyctl(2) will fail for loading PKCS#8 private keys since there is
/usr/lib/modules-load.d/pkcs8.conf:# no automatic module loading for key type parsers.  This entry ensures
/usr/lib/modules-load.d/pkcs8.conf:# that the kernel module pkcs8_key_parser.ko is loaded at boot time.
/usr/lib/modules-load.d/pkcs8.conf:pkcs8_key_parser
/usr/lib/modules-load.d/bluez.conf:crypto_user
/usr/lib/modules-load.d/virtualbox-host-dkms.conf:vboxdrv
/usr/lib/modules-load.d/virtualbox-host-dkms.conf:vboxnetadp
/usr/lib/modules-load.d/virtualbox-host-dkms.conf:vboxnetflt
```

```
29# sudo /lib/systemd/systemd-modules-load
Module 'sg' is built in
```

It looks like` /usr/lib/modules-load.d/cdrecord.conf ` is asking for the sg module.
But why isn't that an issue with mkinitcpio?

**Comment by haraldh** [comment]
Is `modules.builtin` included in your initramfs?

```shell
$ lsinitrd <initrdfile> | grep modules.builtin
```

**Comment by mabod** [comment]
```
37# lsinitrd initramfs-linux-xanmod-lts.img | grep modules.builtin
-rw-r--r--   0 root     root           12 Jan  1  1970 usr/lib/modules/5.10.34-xanmod1-1-lts/modules.builtin.alias.bin
-rw-r--r--   0 root     root        12923 Jan  1  1970 usr/lib/modules/5.10.34-xanmod1-1-lts/modules.builtin.bin
```

**Comment by haraldh** [comment]
> EDIT:
> And `/usr/lib/modules-load.d/cdrecord.conf` is not the issue. I moved this file somehwere else and I still get the warning: `systemd-modules-load[299]: Failed to find module 'sg'`

Well, it might still be in the initramfs.

```shell
# lsinitrd initramfs-linux-xanmod-lts.img | grep modules-load.d
```

**Comment by haraldh** [comment]
also try this:

```shell
# lsinitrd initramfs-linux-xanmod-lts.img usr/lib/modules/5.10.34-xanmod1-1-lts/modules.builtin | grep sg.ko
```

**Comment by mabod** [comment]
```
12# lsinitrd initramfs-linux-xanmod-lts.img -f usr/lib/modules/5.10.34-xanmod1-1-lts/modules.builtin | grep sg.ko
kernel/block/bsg.ko
kernel/drivers/scsi/sg.ko
```

**Comment by haraldh** [comment]
> The PC is still booting and I do not experience any misbehavior. But the message in the log is annoying.
> 
> Can the dependency on the `sg.ko` module be removed from dracut?

The instruction in `module-setup.sh` only installs the kernel module into the initrd image.
Due to `/usr/lib/modules-load.d/cdrecord.conf` being installed into the initrd, `systemd-modules-load` is trying to load it and causes:
```
systemd-modules-load[297]: Failed to find module 'sg'
```

Although `modules.builtin` contains `sg.ko`, so `systemd-modules-load` should only display `Module 'msr' is built in`.

**Comment by haraldh** [comment]
So, you will have to add `rd.shell rd.break` to your kernel command line and run:
```
# /lib/systemd/systemd-modules-load 
# modinfo sg
```

manually to see if the `modinfo` output also reports: 
```
# modinfo sg
name:           sg
filename:       (builtin)
[…]
```

**Comment by haraldh** [comment]
This is likely the culprit: https://github.com/dracutdevs/dracut/pull/1436

**Comment by mabod** [comment]
modinfo does not find the module. See screenshot:

![21-05-04 11-25-23 4256](https://user-images.githubusercontent.com/10911665/116984809-eed1d680-accb-11eb-8fa6-521222d294db.jpg)

**Comment by haraldh** [comment]
> modinfo does not find the module. See screenshot:
> 
> ![21-05-04 11-25-23 4256](https://user-images.githubusercontent.com/10911665/116984809-eed1d680-accb-11eb-8fa6-521222d294db.jpg)

yeah, the fix is PR #1436

**Comment by mabod** [comment]
> This is likely the culprit: #1436

Yes! I tested it and it fixes my issue.
Thanks.
